### PR TITLE
[3805] - Implement sorting and filtering for public courses endpoint

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/u8s9har0/3805-s-linked-to-3789-add-sorting-to-api-public-v1-recruitmentcycles-year-courses
https://trello.com/c/t0UwzFDt/3806-s-linked-to-3789-add-filtering-to-api-public-v1-recruitmentcycles-year-courses
- Dependent on this PR https://github.com/DFE-Digital/teacher-training-api/pull/1534

### Changes proposed in this pull request

- Implements sorting/filtering via the CourseSearchService in `/api/public/v1/:recruitment_cycle_year/courses`

### Guidance to review

-  Fire off the example request below and add your sorting and filtering options in the request.
- Assert the courses returned match provided sorting/filtering params

CURL:

```
curl --location --request GET 'localhost:3001/api/public/v1/recruitment_cycles/2020/courses?sort=name,provider.provider_name&filter[funding_type]=salary' \
--header 'Content-Type: application/json'
```

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by darokel** [comment]
[[S] [Linked to 3789] - Add filtering to /api/public/v1/recruitment_cycles/:year/courses](https://trello.com/c/t0UwzFDt/3806-s-linked-to-3789-add-filtering-to-api-public-v1-recruitmentcycles-year-courses)

**Comment by darokel** [comment]
[[S] [Linked to 3789] Add sorting to /api/public/v1/recruitment_cycles/:year/courses](https://trello.com/c/u8s9har0/3805-s-linked-to-3789-add-sorting-to-api-public-v1-recruitmentcycles-year-courses)

**Review by asmega — APPROVED** [review]

File: spec/controllers/api/public/v1/courses_controller_spec.rb [hunk 1]
```diff
@@ -70,6 +74,46 @@
           expect(provider_id).to eq(provider.id)
         end
       end
+
+      context "with sorting" do
```
Comment (misaka) [inline spec/controllers/api/public/v1/courses_controller_spec.rb:78]: Just noticed that `with includes` above tests how the JSON API renderer works, it's response, etc. This section only tests it calls `CourseSearchService` the right way. Minor point, could be worth aligning.

**Review by misaka — APPROVED** [review]

File: spec/controllers/api/public/v1/courses_controller_spec.rb (hunk 1 above)
Comment (darokel) [reply spec/controllers/api/public/v1/courses_controller_spec.rb:78]: Yeah this was intentional. The `CourseSearchService` is already tested quite extensively. I didn't see the value in duplicating those here in the controller. I am relying on that search class to do its job but asserting it gets the right params to perform that task.

I suppose we could lead with that same assumption for the include scenario but I want to sanity check the include params are being whitelisted properly by the JSON API library we use and that it returns the correct data back.

**Pull request merged**